### ISSUE TITLE
argparse eats initial '=' from short option argument

### ISSUE THREAD

**Issue by netch80** [python/cpython#124305]
# Bug report

### Bug description:

When configured with a short option that requires an argument, and an input commandline has the option argument at the same argv item, argparse deletes initial '=' from it. This behavior is inconsistent with other implementations of the same parsing (at least: getopt, getopt_long from glibc or FreeBSD libc; getopt and optparse from Python standard library).

Actual result: initial '=' is deleted under the described conditions.

Expected result: initial '=' is kept in the collected option value.

Example code:
```python
import argparse
parser = argparse.ArgumentParser()
parser.add_argument('-Q', type=str)
args = parser.parse_args()
print(getattr(args, 'Q', None))
```
Running it:
```
$ ./t_eq_argparse.py -Q22
22
$ ./t_eq_argparse.py -Q=22
22
```
With a separate argument, truncating has no place:
```
$ ./t_eq_argparse.py -Q =22
=22
```
Compare to getopt module:
```
$ ./t_eq_getopt.py -Q22
22
$ ./t_eq_getopt.py -Q=22
=22
```
The same result with libc getopt (Ubuntu 22.04) and optparse module. Test code for getopt and optparse is obvious.

The issue is present with today's Lib/argparse.py from github trunk.

### CPython versions tested on:

3.10, CPython main branch

### Operating systems tested on:

Linux

### Linked PRs
* gh-124332

**Comment by netch80** [comment]
Credits for finding to: discussion at: https://habr.com/ru/companies/timeweb/articles/840200/

**Comment by savannahostrowski** [comment]
While I don't have the historical context to understand why this behaviour differs between optparse and argparse (beyond additional handling of more complex arguments), I feel like this is expected behaviour. Using = as a delimiter between the option and the value is pretty standard.

Changing the behaviour is also likely to be a hugely breaking change at this point, given that many folks rely on this convention.

**Comment by netch80** [comment]
> Using = as a delimiter between the option and the value is pretty standard.

It is not. It could be in another universe, but this is not how other libraries work now.

Forcing '=' between option name and value, if they are in the same argv item, is the matter of fact for _long_ options because 1) they need an explicit name terminator and 2) they allow optional argument case (and this could have been also done because of opportunity by (1)). But this is not how _short_ options work. They are either without an argument or with mandatory argument, and '=' is not added in the latter case.

You may, of course, make the current behavior nailed and  documented... but you'll gather curses from all programmers that face the discrepancy.

**Comment by lsap** [comment]
Let me post more specific link to the discussion: https://habr.com/en/companies/timeweb/articles/840200/comments/

**Comment by serhiy-storchaka** [comment]
It is unfortunate that the behavior of `argparse` differs from that of other parsers in this particular case. There are reasons for this:
* `argparse` supports long options with a single `-` prefix, e.g. `-foo`.
* `argparse` supports abbreviation of options, and it is enabled by default. So `-f` and `-fo` may be abbreviations of long option `-foo`.

As result, `-f` may require `=` if it is an abbreviation of a long option to distinguish the cases when `-fo` is and abbreviation of `-foo` and when it is a short option `-f` with argument `o`.

It does not mean that we should not try to resolve conflict between `argparse` and other parsers. I'll look what can be done.

**Comment by serhiy-storchaka** [comment]
#124332 fixes this issue. The only failed tests are tests added by me in e47ecbd0420528f1f9f282d9e7acfcf586a4caa1 (GH-114180) few months ago. They do not reflect intended behavior, they only check that the current behavior (correct or not) will not change unintentionally. So it is unlikely that this fix will break some user code. This behavior was not documented and not tested.

**Comment by serhiy-storchaka** [comment]
There is a problem here. How can you pass an option with an arbitrary value?

1. `-f "$ARG"`
   This does not work if $ARG starts with `-` (in particularly if it is `--`). `argparse` consider it as a different option and does not allow to use as optional argument value. Eating `--` would have catastrophic consequences -- `argparse` does not work with this.

2. `-f"$ARG"`
   This does not work if $ARG starts with `=` (this bug) or if it is empty.

3. `-f="$ARG"`
   This always works, but will stop to work after fixing this bug.

It is difficult to fix case 1. It would require significant rewriting of some code in the core of `argparse`. It will take time. But until it is fixed, we cannot fix 2 by breaking 3. So, although the fix is ​​ready, it has to wait.

**Comment by netch80** [comment]
> -f "$ARG"
> This does not work if $ARG starts with - (in particularly if it is --).

If so, this is another issue. Iʼm uncertain you consider the same things as I see, so I will elaborate. Sorry if this is a full repetition of your conclusion.

Either we allow mixing of options and positional arguments (relaxed style typical now in Linux) or not (Linux with POSIXLY_CORRECT, or some other implementations when it is the only variant), it should parse input sequence without predefinition of marks like "--". Well, it means a kind of quite trivial, but still a "grammar" that does not interpret a value outside of its context.

Tests with glibc getopt() show it processes this correctly. With a minimal code that just parses "xy:" as the spec and prints results, it shows:

```
$ ./prog -y -- aa -y --
y=--
y=--
arg[0]=aa
```
and:
```
$ POSIXLY_CORRECT=1 ./prog -y -- aa -- -y --
y=--
arg[0]=aa
arg[1]=--
arg[2]=-y
arg[3]=--
```

But, if `--` is when a new input chunk is expected:

```
$ ./prog -y -- aa -- -y --
y=--
arg[0]=aa
arg[1]=-y
arg[2]=--
```

But with argparse, this is immediately broken:
```
parser.add_argument('-x', action = 'count')
parser.add_argument('-y', action = 'append')
parser.add_argument('p', nargs = '?', action = 'append')
```
Run:
```
$ ./py_a.py -y --
usage: py_a.py [-h] [-x] [-y Y] [p]
py_a.py: error: argument -y: expected one argument
```

This reveals a larger problem in argparse: seems it is incorrect in the gory basics.

Unlike argparse, optparse handles this correctly:

```
$ ./py_o.py -y -- aa -y --
opts=<Values at 0x7d6cc5f8cd90: {'x': None, 'y': ['--', '--']}>
args=['aa']
```
Python getopt has another problem that it accepts a positional argument just once:
```
$ ./py_g.py -y -- aa -y --
opts=[('-y', '--')]
args=['aa', '-y', '--']
```
But this is, eghhmm, wrong but a bit less wrong ;\

To sum up: I guess the issue of  misplaced detection of '--' in argparse is the first thing to fix. After it, it should be easier to deal with the original issue.

Attached: source code for the tests here (absolutely trivial).

Also I see that the current release notes for Python suggest getopt and optparse module are "soft deprecated"... IMHO this decision should be reverted until the current issues fixing.

[sources1.zip](https://github.com/user-attachments/files/17518133/sources1.zip)

**Comment by serhiy-storchaka** [comment]
This is a part of the `argparse` design they consider a user error (missed argument for option that requires an argumen) is more likely than passing a value that starts with `-`. It can't be helped, the code hardly depends on the fact that argument values cannot start with `-` (except when they look like negative numbers). And some features (e.g. `nargs='?'` or `nargs='*'`)depends on this convention. To change this, a completely different parsing code should be written, it will not support all `argparse` features, and it cannot be used by default.

This means also that programs which currently use `argparse` will keep their behavior. Only new programs can have different behavior if they explicitly use a new parsing mode.

> Python getopt has another problem that it accepts a positional argument just once:

`getopt.getopt()` implements the old Posix getopt(). Use `getopt.gnu_getopt()` for the GNU getopt() behavior.

I agree that there is nothing wrong in using the `getopt` or `optparse` modules. They are not broken and not outdated. The new shiny `argparse` has too many unsolved problems.
